### PR TITLE
[CI] Refactor tests to reduce CI time.

### PR DESCRIPTION
**Pull request by RAMitchell**
Also xfail failing test from #8307.

### CODE REVIEW

**Comment by RAMitchell** [comment]
CPU python tests 24m:42s->14m:26s.
GPU python tests 5m:27s->1m:54s.

**Comment by RAMitchell** [comment]
Test XGBoost on Windows: 21min->12min

**Comment by hcho3** [comment]
The failing MacOS tests are fixed in #8321

File: tests/python/testing.py [hunk 1]
```diff
@@ -204,7 +204,6 @@ def next(self, input_data: Callable) -> int:
             label=self.y[self.it].copy(),
             weight=self.w[self.it].copy() if self.w else None,
         )
-        gc.collect()  # clear up the copy, see if XGBoost access freed memory.
```
Comment (trivialfis) [inline tests/python/testing.py:207]: is this change related? I put it there along with the "copy" function above to help prevent XGBoost from accessing freed memory.

File: tests/python/test_with_sklearn.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 
 rng = np.random.RandomState(1994)
 
-pytestmark = pytest.mark.skipif(**tm.no_sklearn())
+pytestmark = [pytest.mark.skipif(**tm.no_sklearn()), pytest.mark.timeout(30)]
```
Comment (trivialfis) [inline tests/python/test_with_sklearn.py:15]: Which test might timeout?

File: tests/python/test_with_dask.py [hunk 3]
```diff
@@ -64,8 +66,8 @@ def client(cluster):
         yield dask_client
 
 
-kRows = 1000
-kCols = 10
+kRows = 300
```
Comment (trivialfis) [inline tests/python/test_with_dask.py:69]: 1000 is pretty trivial, right? Is it becoming a bottleneck? If the data is too small dask might put them into a single worker.

**Comment by trivialfis** [comment]
Apologies for the inconvenience, the failing test should be fixed by https://github.com/dmlc/xgboost/pull/8323 .

File: tests/python/testing.py (hunk 1 above)
Comment (RAMitchell) [reply tests/python/testing.py:207]: The repeated calls to gc are the main cause of slowness in external memory tests. Although I wonder if removing this line is causing the segfault in mgpu tests.

File: tests/python/test_with_sklearn.py (hunk 2 above)
Comment (RAMitchell) [reply tests/python/test_with_sklearn.py:15]: Times taken from CI:

```
36.49s call     python/test_with_sklearn.py::test_parameter_tuning
34.58s call     python/test_with_sklearn.py::test_RFECV
```

I made the test under 30s now, but I am placing timeouts to prevent others from adding any slower tests than currently exist.

Comment (RAMitchell) [reply tests/python/test_with_sklearn.py:15]: Ideally we will be targeting <5s for python tests with some few exceptions.

Comment (trivialfis) [reply tests/python/test_with_sklearn.py:15]: Got it, thank you for reducing the time!

File: tests/python/testing.py (hunk 1 above)
Comment (trivialfis) [reply tests/python/testing.py:207]: That would be really interesting to understand why.

File: tests/python/test_with_dask.py (hunk 3 above)
Comment (RAMitchell) [reply tests/python/test_with_dask.py]: Changed this one back. There is not too much difference.

**Comment by RAMitchell** [comment]
I'm still getting intermittent segfaults in the mgpu tests somehow.

**Review by trivialfis — APPROVED** [review]
